File: truevfs-ext/truevfs-ext-insight/src/main/java/net/java/truevfs/ext/insight/I5tStatistics.java
/*
 * Copyright (C) 2005-2012 Schlichtherle IT Services.
 * All rights reserved. Use is subject to license terms.
 */
package net.java.truevfs.ext.insight;

import javax.annotation.concurrent.ThreadSafe;
import javax.management.ObjectName;
import net.java.truevfs.comp.jmx.JmxColleague;
import static net.java.truevfs.comp.jmx.JmxUtils.register;
import net.java.truevfs.ext.insight.stats.FsStatistics;

/**
 * A controller for {@linkplain FsStatistics file system statistics}.
 * 
 * @author Christian Schlichtherle
 */
@ThreadSafe
abstract class I5tStatistics implements JmxColleague {

    private final I5tMediator mediator;
    private final int offset;

    I5tStatistics(final I5tMediator mediator, final int offset) {
        assert 0 <= offset;
        this.mediator = mediator;
        this.offset = offset;
    }

    String getSubject() { return mediator.getSubject(); }

    FsStatistics getStats() { return mediator.getStats(offset); }

    void rotate() { mediator.rotateStats(this); }

    private ObjectName getObjectName() {
        return mediator.nameBuilder(FsStatistics.class)
                .put("subject", getSubject())
                .put("offset", mediator.formatOffset(offset))
                .get();
    }

    abstract I5tStatisticsView newView();

    @Override
    public void start() { register(getObjectName(), newView()); }

    /**
     * Returns a string representation of this object for debugging and logging
     * purposes.
     */
    @Override
    public String toString() {
        return String.format("%s[subject=%s, offset=%d, mediator=%s]",
                getClass().getName(), getSubject(), offset, mediator);
    }
}

File: truevfs-ext/truevfs-ext-insight/src/main/scala/net/java/truevfs/ext/insight/I5tStatistics.scala
/*
 * Copyright (C) 2005-2015 Schlichtherle IT Services.
 * All rights reserved. Use is subject to license terms.
 */
package net.java.truevfs.ext.insight

import javax.annotation.concurrent._
import net.java.truevfs.comp.jmx._
import net.java.truevfs.ext.insight.stats._

/**
 * A controller for [[net.java.truevfs.ext.insight.stats.FsStatistics]].
 *
 * @author Christian Schlichtherle
 */
@ThreadSafe
private abstract class I5tStatistics(mediator: I5tMediator, offset: Int)
extends JmxComponent {

  assert(0 <= offset)

  def subject: String = mediator.subject
  def stats: FsStatistics = mediator stats offset
  def rotate(): Unit = { mediator rotateStats this }

  private def objectName = mediator
    .nameBuilder(classOf[FsStatistics])
    .put("subject", subject)
    .put("offset", mediator formatOffset offset)
    .get

  def newView: I5tStatisticsView

  override def activate(): Unit = { mediator register (objectName, newView) }

  override def toString: String =
    "%s[subject=%s, offset=%d, mediator=%s]" format (getClass.getName, subject, offset, mediator)
}
